File: 킬링캠프/3주 차 - BFS, DFS/practice/Shortest Path in Binary Matrix/Solution.java
import java.util.*;

class Solution {

    private static final int[][] DIR = new int[][]{{-1, 0}, {0, 1}, {0, -1}, {1, 0}, {-1, 1}, {-1, -1}, {1, 1}, {1, -1}}; // 북, 동, 서, 남, 동북, 서북, 동남, 서남

    public int shortestPathBinaryMatrix(int[][] grid) {
        if (grid[0][0] == 1 || grid[grid.length - 1][grid[0].length - 1] == 1) {
            return -1;
        }

        return bfs(grid);
    }

    private int bfs(int[][] grid) {
        int gridWidth = grid.length;
        int gridHeight = grid[0].length;
        boolean[][] visited = new boolean[gridWidth][gridHeight];
        Deque<Node> dq = new ArrayDeque<>();
        dq.offer(new Node(0, 0, 1));
        visited[0][0] = true;

        while (!dq.isEmpty()) {
            Node cur = dq.poll();
            if (cur.row == (gridWidth - 1) && cur.col == (gridHeight - 1)) {
                return cur.seq;
            }

            for (int i = 0; i < 8; i++) {
                Node next = new Node(cur.row + DIR[i][0], cur.col + DIR[i][1], cur.seq + 1);
                if (!isSafe(grid, next) || visited[next.row][next.col]) {
                    continue;
                }

                dq.offer(next);
                visited[next.row][next.col] = true;
            }
        }

        return -1;
    }

    private boolean isSafe(int[][] grid, Node node) {
        int row = node.row;
        int col = node.col;

        return row >= 0 && row < grid.length && col >= 0 && col < grid[row].length && grid[row][col] == 0;
    }
}

class Node {
    int row;
    int col;
    int seq;

    Node(int row, int col, int seq) {
        this.row = row;
        this.col = col;
        this.seq = seq;
    }
}

File: 킬링캠프/3주 차 - BFS, DFS/practice/Shortest Path in Binary Matrix/solution.py
from collections import deque

class Solution:

    DIR = [[-1, 0], [0, 1], [0, -1], [1, 0], [-1, -1], [-1, 1], [1, -1], [1, 1]]

    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        if grid[0][0] == 1:
            return -1
        
        target_row = len(grid) - 1
        target_col = len(grid[0]) - 1
        visited = [[False for _ in range(len(grid[0]))] for _ in range(len(grid))]
        dq = deque()

        dq.append([0, 0, 1])
        visited[0][0] = True

        while dq:
            current_row, current_col, current_step = dq.popleft()
            if current_row == target_row and current_col == target_col:
                return current_step

            for dr, dc in self.DIR:
                next_row = current_row + dr
                next_col = current_col + dc

                if not self.is_safe(next_row, next_col, grid, visited):
                    continue

                dq.append([next_row, next_col, current_step + 1])
                visited[next_row][next_col] = True
        
        return -1
    
    def is_safe(self, row, col, grid, visited):
        return row >= 0 and row < len(grid) and col >= 0 and col < len(grid[row]) and grid[row][col] == 0 and not visited[row][col]
